portfolio/src/main/java/com/google/sps/servlets/VoteServlet.java

```diff
@@ -40,37 +43,51 @@
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String parsedBody = CharStreams.toString(request.getReader());
-
-    JsonParser parser = new JsonParser();
-    JsonElement parsedJson = parser.parse(parsedBody);
-    JsonObject jsonObject = parsedJson.getAsJsonObject();
+    JsonObject jsonObject = UtilityFunctions.stringToJsonObject(parsedBody);
 
     long commentId = Long.parseLong(UtilityFunctions.getFieldFromJsonObject(
         jsonObject, "id", "0"));
-    if (commentId != 0) {
+    long amount = Long.parseLong(UtilityFunctions.getFieldFromJsonObject(
+        jsonObject, "amt", "0"));
+
+    if (commentId != 0 && amount != 0) {
       boolean isUpvote = Boolean.parseBoolean(UtilityFunctions.getFieldFromJsonObject(
-          jsonObject, "isupvote", "true"));
-      changeVoteInDatastore(commentId, isUpvote);
+        jsonObject, "isupvote", "true"));
+      changeVoteInDatastore(commentId, isUpvote, amount);
     }
   }
 
   /* 
    * Changes the number of upvotes/downvotes of the comment IDed by commentId
    * based on whether isUpvote is true or false as well as the net score of the comment
    */
-  private void changeVoteInDatastore(long commentId, boolean isUpvote) {
+  private void changeVoteInDatastore(long commentId, boolean isUpvote, long amount) {
     Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
     KeyFactory keyFactory = datastore.newKeyFactory().setKind("Comment");
     Entity comment = datastore.get(keyFactory.newKey(commentId));
+    
+    String voters = comment.getString("voters");
+    JsonObject obj = UtilityFunctions.stringToJsonObject(voters);
+    String userId = UtilityFunctions.getCurrentUserId();
+
+    if (obj.has(userId) && amount == 1) {
```

Comment:
Since it is not obvious what these branches mean, could you add comments explaining what each branch means? i.e. the first branch is "user already voted upvote/downvote and is trying to vote the same again"